File: linked-list-reversal/reverse-linked-list-ii.py
from linked_list import LinkedList
from linked_list_node import LinkedListNode


def reverse_between(head, left, right):
    prehead = LinkedListNode(0, head)
    curr = prehead

    # phase 1: traverse until left position node
    for i in range(left - 1):  # get to node just before left
        curr = curr.next

    # save that node, then move to left position node
    node_before = curr
    curr = curr.next

    # phase 2: begin reversing for the specified length
    prev = None
    for i in range(right - left + 1):
        next = curr.next
        curr.next = prev
        prev = curr
        curr = next

    # phase 2: connect links
    # connect reversed tail to node-after-right
    node_before.next.next = curr
    # connect node-before-left to reversed head
    node_before.next = prev

    # return head
    return prehead.next

File: linked-list-reversal/reverse-linked-list-ii.js
import LinkedList from "./linked_list.js";
import LinkedListNode from "./linked_list_node.js";

export function reverseBetween(head, left, right) {
	let preHead = new LinkedListNode(0, head);
	let curr = preHead;
	let preLeft = null;

	for (let i = 0; i < left; i++) {
		preLeft = curr;
		curr = curr.next;
	}

	let prev = null;

	for (let i = 0; i < right - left + 1; i++) {
		let next = curr.next;
		curr.next = prev;
		prev = curr;
		curr = next;
	}

	preLeft.next.next = curr;
	preLeft.next = prev;

	return preHead.next;
}
